### DISCUSSION TITLE
keys here compared to other sql databases

### DISCUSSION THREAD

**Discussion by boogerlad** [agentm/project-m36#335] [General]
In SQL databases, I usually see primary keys defined as an autoincrementing integer that starts at 0. However, in this project, I see the primary key as the minimum attributes necessary to uniquely identify a row([for a user relation for example, instead of an id, the email address itself is the key](https://youtu.be/AqheIuSQjeM?t=1659)), or a [text field in the form](https://github.com/agentm/project-m36/blob/master/docs/tutd_tutorial.markdown#relational-expressions) of `$RELATION_NAME#::text` and `$RELATION_NAME$AUTOINCREMENTING_INTEGER`. Why is that? Doesn't that also make foreign key constraints and joins slower?

I've read https://stackoverflow.com/a/338424 and https://stackoverflow.com/a/840182 which would seem that doing things this way is not "right", but I wonder if the design of project m36 invalidates these concerns? Would this be considered as "accidental state"?

**Comment by agentm** [comment]
Indeed, auto-incrementing keys are a pain-point for many DBMSes with Project:M36 being no exception. A lot of discussion surrounding auto-generated primary keys stems from implementation-specific, performance-centric, or practical considerations, but I will step away from that to discuss the theory.

Consider that in the relational algebra that every table has a "natural" primary key since every tuple within a relation must unique (a relation is a set of tuples), so it's not logically possible to have a relation without any "keys". Of course, we can't deny the usefulness of other keys which we can enforce with database constraints.

Project:M36 has deliberately avoided supporting SQL-style "DEFAULT" values for columns (such as auto-incrementing keys), because we haven't found a way to make them mathematically tenable. One problem which you allude to is the unnecessary accidental state introduced by auto-incrementing values (since the value has no inherent meaning to the business at-hand). Another problem is the inherent spooky-action-at-a-distance whereby the value is created by the database and not the client- the client gets to look up the value later... weird.

However, the most concerning issue to me arises from auto-incrementing keys when dealing with merging branches or synchronizing databases. The issue is that key conflicts are unresolvable. If branchA creates a tuple, then branchB creates a tuple in the same relvar (but with the same pk since the branches are unaware of each other), on merge, neither key is more "valid" than the other, but we also cannot change one tuple's key because there might be elements outside the database relying on that key. This conflict would make merging those branches impossible! 

By the way, C.J. Date frowns on the term "primary key" because in the relational algebra, there really nothing more special about one key over another. The term "primary key" exists in SQL-land because we assume that that column will have some index, but the math doesn't require that. SQL doesn't even offer us the courtesy of defaulting to primary keys in join conditions- we still have to type out the complete keys, so what's the point? Oh well, that's just tangential.

Finally, with regards to your performance question, Project:M36's backend handles matters so differently from other databases, it's hard to make apples-to-apples comparisons. For example, Project:M36 takes lazy evaluation to the extreme and does not materialize relvars except on-demand (similarly to Haskell's call-by-need). In addition, just because a unique key appears to be a text value doesn't mean that the database must use that key- there might be implementation-specific, hidden unique keys, hashing, or other tricks happening in the background. Project:M36's reliance on "[data independence](https://github.com/agentm/project-m36/blob/master/docs/commits_in_constant_time.markdown)" opens the door to many more interesting optimizations, many which we have not even thought up yet!

So, what's the solution? I don't claim to have a magic bullet, but UUIDs do seem the address the branch merge conflicts. UUIDs also do not imply any usefulness outside the "key-ness". By that, I mean they cannot be used in an real business logic outside of comparing for equality. My only reservation is that in allowing clients to create the primary key (via UUID) functions, various lower-quality UUID functions may be used. However, the database will still enforce the uniqueness constraint on that attribute, so, in practice, it's a minor consideration.

In the future, we might cook up other schemes to deal better with natural keys. For example, if we have a list of companies with the company name as a key and a company changes its name, we could support schemes to propagate those changes wherever that key is mentioned. After all, the database holds that information- refactoring that information should be its speciality, not a cause for consternation.

Thanks, @boogerlad, for your great questions! Feel free to also join us on irc.libera.chat #project-m36 for live discussions as well!

**Reply by boogerlad** [reply]
> we could support schemes to propagate those changes wherever that key is mentioned

yes, absolutely! a programming language named [lamdu](https://github.com/lamdu/lamdu) does this to seamlessly rename functions and have it available in other natural languages. it's also written in haskell ;)

**Reply by boogerlad** [reply]
Interesting. I tried
```
:importexample cjdate
update s where s#="S2" (s#:="s21")
```
which resulted is `ERR: InclusionDependencyCheckError "s_sp_fk" Nothing`
That makes sense because then `sp` will have invalid data, but I expected it to update the value is `sp` as well (like most SQL databases). Is it not done currently for performance or mathematical reasons?

> In addition, just because a unique key appears to be a text value doesn't mean that the database must use that key

it would be very interesting if there was a uuid behind the value, such that updating only needs to happen once, since all foreign keys are actually to the identifier and not the value

**Reply by agentm** [reply]
Project:M36 implements the semi-standard TutorialD as proposed by C.J. Date. Cascading updates through keys like this has not implemented, but it could be implemented from the userspace interface using a database context function to introspect and identify foreign keys like this and generate additional updates.

**Reply by agentm** [reply]
It's not clear to me that propagating foreign key updates is safe. Is there a good reason that SQL databases support cascading deletes on foreign keys but not cascading updates?

**Reply by boogerlad** [reply]
postgres supports cascading updates

https://www.postgresql.org/docs/current/ddl-constraints.html#DDL-CONSTRAINTS-FK
> Analogous to ON DELETE there is also ON UPDATE which is invoked when a referenced column is changed (updated). The possible actions are the same. In this case, CASCADE means that the updated values of the referenced column(s) should be copied into the referencing row(s).

https://dba.stackexchange.com/questions/254605/is-it-a-good-or-bad-idea-to-use-on-update-cascade-on-delete-cascade-for-foreig#comment565072_254605

> "default behaviour is always (or at least should always be) the most commonly needed behaviour" Not with database systems. Default behavior is usually the safest behavior. That is, the behavior least likely to destroy data.

> I'm not sure about ON UPDATE CASCADE. If you find yourself needing this sort of cascaded update then that is perhaps a "code smell" in your database design. In theory your primary key should be static so changes that need cascading shouldn't need to happen.

it's not a code smell if the value only gets updated once instead of for each reference imo, but I'll let you be the judge on how it gets implemented
